compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp

```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
+    if (!forallOp) {
+      return failure();
+    }
+    // This allows us to assume that the extract/inserts in the loop are
+    // disjoint and makes the application of this pattern safe.
+    if (!forallOpHasMappingType<IREE::Codegen::WorkgroupMappingAttr>(
+            forallOp)) {
+      return failure();
+    }
+    // This pattern only supports forall ops with single
+    // output.
+    SmallVector<Value> forallOutputs(forallOp.getOutputs());
+    if (forallOutputs.size() != 1) {
+      return failure();
+    }
+
+    SmallVector<ReassociationIndices> reIndices =
+        collapseOp.getReassociationIndices();
+    SmallVector<int64_t> expandedDestShape;
+    SmallVector<int64_t> totalInnerSizes;
+    // Get the shape of the outer expand which will be the new destination
+    // of the scf.forall and the total size of inner dimensions per uncollapsed
+    // dimension.
+    if (failed(getExpandedShape(reIndices, collapseOp.getSrcType().getShape(),
+                                insertDest, expandedDestShape,
+                                totalInnerSizes))) {
+      return failure();
+    }
+
+    // Verify that the users of destination are valid to expand and collect all
+    // such users.
+    SmallVector<Operation *> expandableUsers;
+    if (failed(verifyAndCollectExpandableUsers(
+            insertDest, collapseOp.getReassociationIndices(),
+            expandableUsers))) {
+      return failure();
+    }
+
+    // Expand the users of the destination.
+    rewriter.setInsertionPointToStart(forallOp.getBody());
+    expandVerifiedUsers(rewriter, loc, ctx, expandableUsers, totalInnerSizes,
+                        reIndices, forallOp);
+    rewriter.setInsertionPoint(forallOp);
+
+    Operation *outOp = forallOutputs[0].getDefiningOp();
+    if (!outOp) {
+      return failure();
+    }
+
+    // Create the expand -> new scf.forall -> collapse chain.
+    Type expandedDestType = RankedTensorType::get(
+        expandedDestShape,
+        cast<ShapedType>(outOp->getResult(0).getType()).getElementType());
+    auto expandedDest = rewriter.create<tensor::ExpandShapeOp>(
+        loc, expandedDestType, outOp->getResult(0), reIndices);
+
+    scf::ForallOp newForallOp = rewriter.create<scf::ForallOp>(
+        loc, forallOp.getMixedLowerBound(), forallOp.getMixedUpperBound(),
+        forallOp.getMixedStep(), ValueRange{expandedDest},
+        forallOp.getMappingAttr());
+
+    auto collapsedResultOp = rewriter.create<tensor::CollapseShapeOp>(
+        loc, cast<ShapedType>(forallOp->getResult(0).getType()),
+        newForallOp->getResult(0), reIndices);
+
+    // Merge the old scf.forall block which has the expanded users into the new
+    // scf.forall which has the expanded destination.
+    SmallVector<Value> argReplacements(newForallOp.getInductionVars());
+    for (auto forallIterArg : newForallOp.getRegionIterArgs()) {
+      argReplacements.push_back(forallIterArg);
+    }
```

Comment:
nit: I think this can be
```
argReplacements.append(newForallOp.getRegionIterArgs().begin(),
                       newForallOp.getRegionIterArgs().end());
```